File: ruby/lib/enolib/elements/missing/missing_section.rb
# frozen_string_literal: true

module Enolib
  class MissingSection < MissingElementBase
    def element(key = nil)
      MissingSectionElement.new(key, self)
    end

    def elements
      []
    end

    def embed(key = nil)
      MissingEmbed.new(key, self)
    end
    
    def embeds(_key = nil)
      []
    end
    
    def field(key = nil)
      MissingField.new(key, self)
    end
    
    def fields(_key = nil)
      []
    end

    def flag(key = nil)
      MissingFlag.new(key, self)
    end
    
    def flags(_key = nil)
      []
    end

    def optional_element(_key = nil)
      nil
    end

    def optional_embed(_key = nil)
      nil
    end
    
    def optional_field(_key = nil)
      nil
    end
    
    def optional_flag(_key = nil)
      nil
    end

    def optional_section(_key = nil)
      nil
    end

    def required_element(_key = nil)
      @parent._missing_error(self)
    end

    def required_embed(_key = nil)
      @parent._missing_error(self)
    end
    
    def required_field(_key = nil)
      @parent._missing_error(self)
    end
  
    def required_flag(_key = nil)
      @parent._missing_error(self)
    end

    def required_section(_key = nil)
      @parent._missing_error(self)
    end

    def section(key = nil)
      MissingSection.new(key, self)
    end

    def sections(_key = nil)
      []
    end

    def to_s
      if @key
        "#<Enolib::MissingSection key=#{@key}>"
      else
        '#<Enolib::MissingSection>'
      end
    end
  end
end

File: python/enolib/elements/missing/missing_section.py
from .missing_element_base import MissingElementBase
from . import missing_embed
from . import missing_field
from . import missing_flag
from . import missing_section_element

class MissingSection(MissingElementBase):
    def __repr__(self):
        if self._key:
            return f"<class MissingSection key={self._key}>"

        return '<class MissingSection>'

    def element(self, key=None):
        return missing_section_element.MissingSectionElement(key, self)

    def elements(self, _key=None):
        return []

    def embed(self, key=None):
        return missing_embed.MissingEmbed(key, self)

    def embeds(self, _key=None):
        return []
        
    def field(self, key=None):
        return missing_field.MissingField(key, self)

    def fields(self, _key=None):
        return []
    
    def flag(self, key=None):
        return missing_flag.MissingFlag(key, self)
        
    def flags(self, _key=None):
        return []
    
    def optional_element(self, _key=None):
        return None
        
    def optional_embed(self, _key=None):
        return None

    def optional_flag(self, _key=None):
        return None

    def optional_field(self, _key=None):
        return None

    def optional_section(self, _key=None):
        return None

    def required_element(self, _key=None):
        self._parent._missing_error(self)

    def required_embed(self, _key=None):
        self._parent._missing_error(self)

    def required_field(self, _key=None):
        self._parent._missing_error(self)

    def required_flag(self, _key=None):
        self._parent._missing_error(self)

    def required_section(self, _key=None):
        self._parent._missing_error(self)

    def section(self, key=None):
        return MissingSection(key, self)

    def sections(self, _key=None):
        return []
